test_pydocgui.py has a timeout failure on test_get_pydoc

<!--- **PLEASE READ:** When submitting here, please ensure you've completed the following checklist and checked the boxes to confirm. Issue reports without it may be closed. Thanks! --->

### Issue Report Checklist

* [x] Searched the [issues page](https://github.com/spyder-ide/spyder/issues?q=is%3Aissue) for similar reports
* [x] Read the relevant sections of the [Spyder Troubleshooting Guide](https://github.com/spyder-ide/spyder/wiki/Troubleshooting-Guide-and-FAQ) and followed its advice
* [x] Reproduced the issue after updating with ``conda update spyder`` (or ``pip``, if not using Anaconda)
* [ ] Could not reproduce inside ``jupyter qtconsole`` (if console-related)
* [x] Tried basic troubleshooting (if a bug/error)
    * [ ] Restarted Spyder
    * [ ] Reset preferences with ``spyder --reset``
    * [ ] Reinstalled the latest version of [Anaconda](https://www.anaconda.com/download/)
    * [ ] Tried the other applicable steps from the Troubleshooting Guide
* [x] Completed the **Problem Description**, **Steps to Reproduce** and **Version** sections below


## Problem Description

I tried running the package tests (using runtest.py). I get several errors, but it probably makes more sense to deal with each one separately, so that's what I'm doing.  (See #17057 for the first error; this is the second one.)

This issue is about `spyder/plugins/onlinehelp/tests/test_pydocgui.py` in 5.2.0.  All three tests in this file give errors.  It occurs with Python 3.9 and 3.10; I haven't been able to test with other Python versions.

### What steps reproduce the problem?

Here's an attempt at running just these tests; the output from `runtests.py` is essentially identical.  I ran it inside a chroot with only the listed requirements (plus git, which was needed but not listed) installed.  This gave me three errors; I'm snipping most of the second two, as they are very similar to the first one:
```
(sid):~/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1 $ xvfb-run -a -s "-screen 0 1024x768x24 +extension GLX" python3.9 -m pytest spyder/plugins/onlinehelp/tests/test_pydocgui.py
============================= test session starts ==============================
platform linux -- Python 3.9.9, pytest-6.2.5, py-1.10.0, pluggy-0.13.0
PyQt5 5.15.6 -- Qt runtime 5.15.2 -- Qt compiled 5.15.2
rootdir: /home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1, configfile: pytest.ini
plugins: typeguard-2.11.1, flaky-3.7.0, xvfb-2.0.0, mock-3.6.1, cov-3.0.0, order-1.0.0, qt-4.0.2, lazy-fixture-0.6.3
collected 3 items                                                              

spyder/plugins/onlinehelp/tests/test_pydocgui.py EE                      [100%] [100%] [100%] [100%]E [100%]

==================================== ERRORS ====================================
_____________________ ERROR at setup of test_pydocbrowser ______________________
SETUP ERROR: Exceptions caught in Qt event loop:
________________________________________________________________________________
Traceback (most recent call last):
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 268, in _continue_initialization
    self.go_home()
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 444, in go_home
    self.set_url(self.home_url)
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 425, in set_url
    self.go_to(url)
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 436, in go_to
    self.webview.load(url)
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/widgets/browser.py", line 388, in load
    self.focusProxy().installEventFilter(self)
AttributeError: 'NoneType' object has no attribute 'installEventFilter'
________________________________________________________________________________
---------------------------- Captured stderr setup -----------------------------
Exceptions caught in Qt event loop:
________________________________________________________________________________
Traceback (most recent call last):
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 268, in _continue_initialization
    self.go_home()
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 444, in go_home
    self.set_url(self.home_url)
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 425, in set_url
    self.go_to(url)
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/widgets.py", line 436, in go_to
    self.webview.load(url)
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/widgets/browser.py", line 388, in load
    self.focusProxy().installEventFilter(self)
AttributeError: 'NoneType' object has no attribute 'installEventFilter'
________________________________________________________________________________
____________________ ERROR at setup of test_get_pydoc[lib0] ____________________
SETUP ERROR: Exceptions caught in Qt event loop:
[...snipped...]
AttributeError: 'NoneType' object has no attribute 'installEventFilter'
________________________________________________________________________________
____________________ ERROR at setup of test_get_pydoc[lib1] ____________________
SETUP ERROR: Exceptions caught in Qt event loop:
[...snipped...]
  File "/home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/widgets/browser.py", line 388, in load
    self.focusProxy().installEventFilter(self)
AttributeError: 'NoneType' object has no attribute 'installEventFilter'
________________________________________________________________________________

	[<TracebackEntry /usr/lib/python3/dist-packages/pytestqt/plugin.py:144>, <TracebackEntry /usr/lib/python3/dist-packages/pytestqt/exceptions.py:68>]

===End Flaky Test Report===
=========================== short test summary info ============================
ERROR spyder/plugins/onlinehelp/tests/test_pydocgui.py::test_pydocbrowser
ERROR spyder/plugins/onlinehelp/tests/test_pydocgui.py::test_get_pydoc[lib0]
ERROR spyder/plugins/onlinehelp/tests/test_pydocgui.py::test_get_pydoc[lib1]
============================== 3 errors in 13.13s ==============================
```

I then ran it on my regular (non-chroot) system, with lots more packages installed, and got a very different result: this time, the first two tests passed, but the second failed on a timeout error:
```
euler:~/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1 $ xvfb-run -a -s "-screen 0 1024x768x24 +extension GLX" python3.9 -m pytest spyder/plugins/onlinehelp/tests/test_pydocgui.py 
============================= test session starts ==============================
platform linux -- Python 3.9.9, pytest-6.2.5, py-1.10.0, pluggy-0.13.0
PyQt5 5.15.6 -- Qt runtime 5.15.2 -- Qt compiled 5.15.2
rootdir: /home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1, configfile: pytest.ini
plugins: typeguard-2.11.1, lazy-fixture-0.6.3, hypothesis-5.43.3, order-1.0.0, anyio-3.4.0, xvfb-2.0.0, mock-3.6.1, xdist-2.4.0, timeout-2.0.1, cov-3.0.0, forked-1.3.0, qt-4.0.2, dependency-0.5.1, flaky-3.7.0
collected 3 items                                                              

spyder/plugins/onlinehelp/tests/test_pydocgui.py ..                      [100%] [100%] [100%] [100%]F [100%]

=================================== FAILURES ===================================
_____________________________ test_get_pydoc[lib1] _____________________________

pydocbrowser = (<pytestqt.qtbot.QtBot object at 0x7f2c2e0db520>, <spyder.plugins.onlinehelp.widgets.PydocBrowser object at 0x7f2c2e0c4d30>)
lib = ('numpy.compat', 'numpy.compat', 2)

    @flaky(max_runs=5)
    @pytest.mark.parametrize(
        "lib", [('str', 'class str', 1),
                ('numpy.compat', 'numpy.compat', 2)
                ])
    @pytest.mark.skipif(
        sys.platform == 'darwin', reason="Does not work on Mac")
    def test_get_pydoc(pydocbrowser, lib):
        """
        Go to the documentation by url.
        Regression test for spyder-ide/spyder#10740
        """
        qtbot, browser = pydocbrowser
        element, doc, matches = lib
    
        webview = browser.webview
        element_url = browser.text_to_url(element)
        with qtbot.waitSignal(webview.loadFinished):
            browser.set_url(element_url)
    
        # Check number of matches. In Python 2 are 3 matches instead
        # of 2 for numpy.compat
>       qtbot.waitUntil(
            lambda: webview.get_number_matches(doc) in [matches, matches + 1])
E       pytestqt.exceptions.TimeoutError: waitUntil timed out in 5000 milliseconds

spyder/plugins/onlinehelp/tests/test_pydocgui.py:68: TimeoutError
=============================== warnings summary ===============================
../../../../../../../usr/lib/python3/dist-packages/qdarkstyle/__init__.py:207
  /usr/lib/python3/dist-packages/qdarkstyle/__init__.py:207: DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead
    _logger.warn("No QCoreApplication instance found. "

spyder/plugins/onlinehelp/tests/test_pydocgui.py::test_pydocbrowser
  /usr/lib/python3/dist-packages/nbconvert/exporters/html.py:110: DeprecationWarning: 'contextfilter' is renamed to 'pass_context', the old name will be removed in Jinja 3.1.
    def markdown2html(self, context, source):

-- Docs: https://docs.pytest.org/en/stable/warnings.html
===Flaky Test Report===

test_get_pydoc[lib0] passed 1 out of the required 1 times. Success!
test_get_pydoc[lib1] failed (4 runs remaining out of 5).
	<class 'pytestqt.exceptions.TimeoutError'>
	waitUntil timed out in 5000 milliseconds
	[<TracebackEntry /home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/tests/test_pydocgui.py:68>]
test_get_pydoc[lib1] failed (3 runs remaining out of 5).
	<class 'pytestqt.exceptions.TimeoutError'>
	waitUntil timed out in 5000 milliseconds
	[<TracebackEntry /home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/tests/test_pydocgui.py:68>]
test_get_pydoc[lib1] failed (2 runs remaining out of 5).
	<class 'pytestqt.exceptions.TimeoutError'>
	waitUntil timed out in 5000 milliseconds
	[<TracebackEntry /home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/tests/test_pydocgui.py:68>]
test_get_pydoc[lib1] failed (1 runs remaining out of 5).
	<class 'pytestqt.exceptions.TimeoutError'>
	waitUntil timed out in 5000 milliseconds
	[<TracebackEntry /home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/tests/test_pydocgui.py:68>]
test_get_pydoc[lib1] failed; it passed 0 out of the required 1 times.
	<class 'pytestqt.exceptions.TimeoutError'>
	waitUntil timed out in 5000 milliseconds
	[<TracebackEntry /home/jdg/debian/spyder-packages/spyder/build-area/spyder-5.2.0+dfsg1/spyder/plugins/onlinehelp/tests/test_pydocgui.py:68>]

===End Flaky Test Report===
=========================== short test summary info ============================
FAILED spyder/plugins/onlinehelp/tests/test_pydocgui.py::test_get_pydoc[lib1]
=================== 1 failed, 2 passed, 2 warnings in 34.59s ===================
Release of profile requested but WebEnginePage still not deleted. Expect troubles !
```
So it would seem that there is a missing dependency causing the `'NoneType' object has no attribute 'installEventFilter'` errors, but I don't know what it is.  Do you have any idea?

There's a separate issue around the timeout, but that may or may not be a problem.

### What is the expected output? What do you see instead?

I expect the tests to pass ;-)

## Versions

* Spyder version: 5.2.0
* Python version: 3.9 and 3.10
* Qt version: 5.15.2
* PyQt version: 5.15.6+dfsg-1+b1
* Operating System name/version: Debian unstable and testing

### Dependencies

Erm, easier said than done as I'm not yet running Spyder 5.2.0 live...

--- a/spyder/plugins/console/tests/test_plugin.py
+++ b/spyder/plugins/console/tests/test_plugin.py
@@ -90,6 +90,7 @@ def test_completions(console_plugin, qtbot):
     assert u'import' in shell.toPlainText()
 
 
+@flaky(max_runs=20)
 def test_handle_exception(console_plugin, mocker):
     """Test that error dialog is called."""
     widget = console_plugin.get_widget()

--- a/spyder/plugins/editor/panels/tests/test_scrollflag.py
+++ b/spyder/plugins/editor/panels/tests/test_scrollflag.py
@@ -73,7 +73,6 @@ line6: Found Results
 # ---------------------------------------------------------------------------
 # Tests
 # ---------------------------------------------------------------------------
-
 def test_enabled(editor_bot):
     """"Test that enabling and disabling the srollflagarea panel make
     it visible or invisible depending on the case."""
@@ -148,7 +147,6 @@ def test_flag_painting(editor_bot, qtbot):
         editor.setTextCursor(cursor)
 
 
-@pytest.mark.skipif(running_in_ci(), reason="Fails on CIs")
 def test_range_indicator_visible_on_hover_only(editor_bot, qtbot):
     """Test that the slider range indicator is visible only when hovering
     over the scrollflag area when the editor vertical scrollbar is visible.
@@ -185,6 +183,7 @@ def test_range_indicator_visible_on_hover_only(editor_bot, qtbot):
     x = int(sfa.width()/2)
     y = int(sfa.height()/2)
     qtbot.mouseMove(sfa, pos=QPoint(x, y), delay=-1)
+    qtbot.wait(500)
 
     assert sfa._range_indicator_is_visible is True
 
@@ -196,7 +195,6 @@ def test_range_indicator_visible_on_hover_only(editor_bot, qtbot):
     qtbot.waitUntil(lambda: not sfa._range_indicator_is_visible)
 
 
-@pytest.mark.skipif(running_in_ci(), reason="Fails on CIs")
 def test_range_indicator_alt_modifier_response(editor_bot, qtbot):
     """Test that the slider range indicator is visible while the alt key is
     held down while the cursor is over the editor, but outside of the
@@ -213,8 +211,8 @@ def test_range_indicator_alt_modifier_response(editor_bot, qtbot):
 
     # Set a long text in the editor and assert that the slider is visible.
     editor.set_text(long_code)
-    editor.resize(600, 150)
     qtbot.waitUntil(lambda: sfa.slider)
+    qtbot.wait(500)
 
     # Set the cursor position to the center of the editor.
     w = editor.width()
@@ -269,4 +267,3 @@ def test_range_indicator_alt_modifier_response(editor_bot, qtbot):
 
 if __name__ == "__main__":  # pragma: no cover
     pytest.main([os.path.basename(__file__)])
-    # pytest.main()

--- a/spyder/plugins/ipythonconsole/plugin.py
+++ b/spyder/plugins/ipythonconsole/plugin.py
@@ -410,7 +410,8 @@ class IPythonConsole(SpyderDockablePlugin):
         Remove std files left by previous Spyder instances.
 
         This is only required on Windows because we can't
-        clean up std files while Spyder is running on it.
+        clean up std files while Spyder is running on that
+        platform.
         """
         if os.name == 'nt':
             tmpdir = get_temp_dir()

--- a/spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py
+++ b/spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py
@@ -12,6 +12,7 @@ Tests for the IPython console plugin.
 
 # Standard library imports
 import codecs
+import glob
 import os
 import os.path as osp
 import shutil
@@ -162,7 +163,8 @@ def ipyconsole(qtbot, request, tmpdir):
     is_cython = True if cython_client else False
 
     # Use an external interpreter if requested
-    external_interpreter = request.node.get_closest_marker('external_interpreter')
+    external_interpreter = request.node.get_closest_marker(
+        'external_interpreter')
     if external_interpreter:
         configuration.set('main_interpreter', 'default', False)
         configuration.set('main_interpreter', 'executable', sys.executable)
@@ -173,7 +175,6 @@ def ipyconsole(qtbot, request, tmpdir):
     # Use the test environment interpreter if requested
     test_environment_interpreter = request.node.get_closest_marker(
         'test_environment_interpreter')
-
     if test_environment_interpreter:
         configuration.set('main_interpreter', 'default', False)
         configuration.set(
@@ -1372,17 +1373,31 @@ def test_kernel_crash(ipyconsole, qtbot):
 
 
 @flaky(max_runs=3)
-@pytest.mark.skipif(not os.name == 'nt', reason="Only works on Windows")
+@pytest.mark.skipif(not os.name == 'nt', reason="Only necessary on Windows")
 def test_remove_old_std_files(ipyconsole, qtbot):
     """Test that we are removing old std files."""
-    # Create empty stderr file in our temp dir to see
-    # if it's removed correctly.
+    shell = ipyconsole.get_current_shellwidget()
+    qtbot.waitUntil(lambda: shell._prompt_html is not None,
+                    timeout=SHELL_TIMEOUT)
+
+    # Create empty std files in our temp dir to see if they are removed
+    # correctly.
     tmpdir = get_temp_dir()
     open(osp.join(tmpdir, 'foo.stderr'), 'a').close()
+    open(osp.join(tmpdir, 'foo.stdout'), 'a').close()
 
-    # Assert that only that file is removed
+    # Assert that only old std files are removed
     ipyconsole._remove_old_std_files()
     assert not osp.isfile(osp.join(tmpdir, 'foo.stderr'))
+    assert not osp.isfile(osp.join(tmpdir, 'foo.stdout'))
+
+    # The current kernel std files should be present
+    for fname in glob.glob(osp.join(tmpdir, '*')):
+        assert osp.basename(fname).startswith('kernel')
+        assert any(
+            [osp.basename(fname).endswith(ext)
+             for ext in ('.stderr', '.stdout', '.fault')]
+        )
 
 
 @flaky(max_runs=10)

--- a/spyder/plugins/ipythonconsole/utils/stdfile.py
+++ b/spyder/plugins/ipythonconsole/utils/stdfile.py
@@ -5,9 +5,11 @@
 # Distributed under the terms of the MIT License
 # (see spyder/__init__.py for details)
 # -----------------------------------------------------------------------------
+
 """
 Class to control a file where stanbdard output can be written.
 """
+
 # Standard library imports.
 import codecs
 import os
@@ -17,7 +19,7 @@ from spyder.py3compat import to_text_string
 from spyder.utils.encoding import get_coding
 
 
-class StdFile():
+class StdFile:
     def __init__(self, filename):
         self.filename = filename
         self._mtime = 0

--- a/spyder/plugins/ipythonconsole/widgets/main_widget.py
+++ b/spyder/plugins/ipythonconsole/widgets/main_widget.py
@@ -1073,6 +1073,7 @@ class IPythonConsoleWidget(PluginMainWidget):
         show_elapsed_time = self.get_conf('show_elapsed_time')
         reset_warning = self.get_conf('show_reset_namespace_warning')
         ask_before_restart = self.get_conf('ask_before_restart')
+        std_dir = self._test_dir if self._test_dir else None
         client = ClientWidget(self,
                               id_=client_id,
                               given_name=given_name,
@@ -1092,7 +1093,7 @@ class IPythonConsoleWidget(PluginMainWidget):
                               css_path=self.css_path,
                               configuration=self.CONFIGURATION,
                               handlers=self.registered_spyder_kernel_handlers,
-                              std_dir=self._test_dir)
+                              std_dir=std_dir)
 
         # Create kernel client
         kernel_client = QtKernelClient(connection_file=connection_file)
@@ -1479,6 +1480,7 @@ class IPythonConsoleWidget(PluginMainWidget):
         reset_warning = self.get_conf('show_reset_namespace_warning')
         ask_before_restart = self.get_conf('ask_before_restart')
         ask_before_closing = self.get_conf('ask_before_closing')
+        std_dir = self._test_dir if self._test_dir else None
         client = ClientWidget(self, id_=client_id,
                               history_filename=get_conf_path('history.py'),
                               config_options=self.config_options(),
@@ -1498,7 +1500,7 @@ class IPythonConsoleWidget(PluginMainWidget):
                               css_path=self.css_path,
                               configuration=self.CONFIGURATION,
                               handlers=self.registered_spyder_kernel_handlers,
-                              std_dir=self._test_dir)
+                              std_dir=std_dir)
 
         self.add_tab(
             client, name=client.get_name(), filename=filename,

--- a/spyder/plugins/maininterpreter/tests/test_confpage.py
+++ b/spyder/plugins/maininterpreter/tests/test_confpage.py
@@ -9,6 +9,9 @@
 # Standard library imports
 import time
 
+# Third-party imports
+import pytest
+
 # Local imports
 from spyder.api.plugins import Plugins
 from spyder.api.plugin_registration.registry import PLUGIN_REGISTRY
@@ -23,11 +26,15 @@ from spyder.utils.pyenv import get_list_pyenv_envs
 # We're also recording the time needed to get them to compare it with the
 # loading time of that config page.
 t0 = time.time()
-get_list_conda_envs()
-get_list_pyenv_envs()
+conda_envs = get_list_conda_envs()
+pyenv_envs = get_list_pyenv_envs()
 GET_ENVS_TIME = time.time() - t0
 
 
+@pytest.mark.skipif(
+    len(conda_envs) == 0 and len(pyenv_envs) == 0,
+    reason="Makes no sense if conda and pyenv are not installed"
+)
 def test_load_time(qtbot):
     # Create Preferences dialog
     main = MainWindowMock()

--- a/spyder/plugins/onlinehelp/tests/test_pydocgui.py
+++ b/spyder/plugins/onlinehelp/tests/test_pydocgui.py
@@ -12,6 +12,8 @@ import sys
 from unittest.mock import MagicMock
 
 # Test library imports
+import numpy as np
+from numpy.lib import NumpyVersion
 import pytest
 from flaky import flaky
 
@@ -27,35 +29,32 @@ def pydocbrowser(qtbot):
     widget = PydocBrowser(parent=None, plugin=plugin_mock, name='pydoc')
     widget._setup()
     widget.setup()
+    widget.resize(640, 480)
+    widget.show()
 
     with qtbot.waitSignal(widget.sig_load_finished, timeout=6000):
         widget.initialize()
 
     qtbot.addWidget(widget)
-    return qtbot, widget
-
-
-@pytest.mark.skipif(
-    sys.platform == 'darwin', reason="Does not work on Mac")
-def test_pydocbrowser(pydocbrowser):
-    """Run Pydoc Browser."""
-    qtbot, browser = pydocbrowser
-    assert browser
+    return widget
 
 
 @flaky(max_runs=5)
 @pytest.mark.parametrize(
-    "lib", [('str', 'class str', 1),
-            ('numpy.compat', 'numpy.compat', 2)
-            ])
+    "lib",
+    [('str', 'class str', [0, 1]), ('numpy.testing', 'numpy.testing', [5, 10])]
+)
 @pytest.mark.skipif(
-    sys.platform == 'darwin', reason="Does not work on Mac")
-def test_get_pydoc(pydocbrowser, lib):
+    (sys.platform == 'darwin' or
+     NumpyVersion(np.__version__) < NumpyVersion('1.21.0')),
+    reason="Fails on Mac and older versions of Numpy"
+)
+def test_get_pydoc(pydocbrowser, qtbot, lib):
     """
     Go to the documentation by url.
     Regression test for spyder-ide/spyder#10740
     """
-    qtbot, browser = pydocbrowser
+    browser = pydocbrowser
     element, doc, matches = lib
 
     webview = browser.webview
@@ -63,10 +62,8 @@ def test_get_pydoc(pydocbrowser, lib):
     with qtbot.waitSignal(webview.loadFinished):
         browser.set_url(element_url)
 
-    # Check number of matches. In Python 2 are 3 matches instead
-    # of 2 for numpy.compat
-    qtbot.waitUntil(
-        lambda: webview.get_number_matches(doc) in [matches, matches + 1])
+    expected_range = list(range(matches[0], matches[1]))
+    qtbot.waitUntil(lambda: webview.get_number_matches(doc) in expected_range)
 
 
 if __name__ == "__main__":

--- a/spyder/plugins/variableexplorer/widgets/arrayeditor.py
+++ b/spyder/plugins/variableexplorer/widgets/arrayeditor.py
@@ -152,20 +152,21 @@ class ArrayModel(QAbstractTableModel):
         self.total_cols = self._data.shape[1]
         size = self.total_rows * self.total_cols
 
-        try:
-            self.vmin = np.nanmin(self.color_func(data))
-            self.vmax = np.nanmax(self.color_func(data))
-            if self.vmax == self.vmin:
-                self.vmin -= 1
-            self.hue0 = huerange[0]
-            self.dhue = huerange[1]-huerange[0]
-            self.bgcolor_enabled = True
-        except (AttributeError, TypeError, ValueError):
-            self.vmin = None
-            self.vmax = None
-            self.hue0 = None
-            self.dhue = None
-            self.bgcolor_enabled = False
+        if not self._data.dtype.name == 'object':
+            try:
+                self.vmin = np.nanmin(self.color_func(data))
+                self.vmax = np.nanmax(self.color_func(data))
+                if self.vmax == self.vmin:
+                    self.vmin -= 1
+                self.hue0 = huerange[0]
+                self.dhue = huerange[1]-huerange[0]
+                self.bgcolor_enabled = True
+            except (AttributeError, TypeError, ValueError):
+                self.vmin = None
+                self.vmax = None
+                self.hue0 = None
+                self.dhue = None
+                self.bgcolor_enabled = False
 
         # Array with infinite values cannot display background colors and
         # crashes. See: spyder-ide/spyder#8093

--- a/spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py
+++ b/spyder/plugins/variableexplorer/widgets/tests/test_arrayeditor.py
@@ -16,14 +16,16 @@ import sys
 from unittest.mock import Mock, ANY
 
 # Third party imports
+from flaky import flaky
 import numpy as np
 from numpy.testing import assert_array_equal
 import pytest
 from qtpy.QtCore import Qt
-from flaky import flaky
+from scipy.io import loadmat
 
 # Local imports
-from spyder.plugins.variableexplorer.widgets.arrayeditor import ArrayEditor, ArrayModel
+from spyder.plugins.variableexplorer.widgets.arrayeditor import (
+    ArrayEditor, ArrayModel)
 
 
 # =============================================================================
@@ -44,12 +46,16 @@ def launch_arrayeditor(data, title="", xlabels=None, ylabels=None):
     return dlg.get_value()
 
 
-def setup_arrayeditor(qbot, data, title="", xlabels=None, ylabels=None):
+# =============================================================================
+# Fixtures
+# =============================================================================
+@pytest.fixture
+def setup_arrayeditor(qtbot, data):
     """Setups an arrayeditor."""
     dlg = ArrayEditor()
-    dlg.setup_and_check(data, title, xlabels=xlabels, ylabels=ylabels)
+    dlg.setup_and_check(data)
     dlg.show()
-    qbot.addWidget(dlg)
+    qtbot.addWidget(dlg)
     return dlg
 
 
@@ -62,49 +68,61 @@ def test_object_arrays(qtbot):
     assert_array_equal(arr, launch_arrayeditor(arr, "object array"))
 
 
-def test_object_arrays_display(qtbot):
+@pytest.mark.parametrize(
+    'data',
+    [np.array([[np.array([1, 2])], 2], dtype=object)],
+)
+def test_object_arrays_display(setup_arrayeditor):
     """
     Test that value_to_display is being used to display the values of
     object arrays.
     """
-    arr = np.array([[np.array([1, 2])], 2], dtype=object)
-    dlg = setup_arrayeditor(qtbot, arr)
+    dlg = setup_arrayeditor
     idx = dlg.arraywidget.model.index(0, 0)
     assert u'[Numpy array]' == dlg.arraywidget.model.data(idx)
 
 
-@pytest.mark.skipif(os.name == 'nt', reason="Segfaults on Windows")
-def test_attribute_errors(qtbot):
+@pytest.mark.parametrize(
+    'data',
+    [loadmat(os.path.join(HERE, 'issue_11216.mat'))['S']],
+)
+def test_attribute_errors(setup_arrayeditor):
     """
     Verify that we don't get a AttributeError for certain structured arrays.
 
     Fixes spyder-ide/spyder#11216 .
     """
-    from scipy.io import loadmat
+    dlg = setup_arrayeditor
     data = loadmat(os.path.join(HERE, 'issue_11216.mat'))
-    dlg = setup_arrayeditor(qtbot, data['S'])
     contents = dlg.arraywidget.model.get_value(dlg.arraywidget.model.index(0, 0))
     assert_array_equal(contents, data['S'][0][0][0])
 
 
-def test_type_errors(qtbot):
+@pytest.mark.parametrize(
+    'data',
+    [np.ones(2, dtype=[('X', 'f8', (2,10)), ('S', 'S10')])],
+)
+def test_type_errors(setup_arrayeditor, qtbot):
     """
     Verify that we don't get a TypeError for certain structured arrays.
 
     Fixes spyder-ide/spyder#5254.
     """
-    arr = np.ones(2, dtype=[('X', 'f8', (2,10)), ('S', 'S10')])
-    dlg = setup_arrayeditor(qtbot, arr)
+    dlg = setup_arrayeditor
     qtbot.keyClick(dlg.arraywidget.view, Qt.Key_Down, modifier=Qt.ShiftModifier)
     contents = dlg.arraywidget.model.get_value(dlg.arraywidget.model.index(0, 0))
     assert_array_equal(contents, np.ones(10))
 
 
-@pytest.mark.skipif(sys.platform == 'darwin', reason="Fails on Mac")
-def test_arrayeditor_format(qtbot):
+@pytest.mark.skipif(not sys.platform.startswith('linux'),
+                    reason="Only works on Linux")
+@pytest.mark.parametrize(
+    'data',
+    [np.array([1, 2, 3], dtype=np.float32)],
+)
+def test_arrayeditor_format(setup_arrayeditor, qtbot):
     """Changes the format of the array and validates its selected content."""
-    arr = np.array([1, 2, 3], dtype=np.float32)
-    dlg = setup_arrayeditor(qtbot, arr, "test array float32")
+    dlg = setup_arrayeditor
     qtbot.keyClick(dlg.arraywidget.view, Qt.Key_Down, modifier=Qt.ShiftModifier)
     qtbot.keyClick(dlg.arraywidget.view, Qt.Key_Down, modifier=Qt.ShiftModifier)
     contents = dlg.arraywidget.view._sel_to_text(dlg.arraywidget.view.selectedIndexes())

--- a/spyder/plugins/workingdirectory/tests/test_workingdirectory.py
+++ b/spyder/plugins/workingdirectory/tests/test_workingdirectory.py
@@ -19,7 +19,7 @@ NEW_DIR = 'new_workingdir'
 
 
 @pytest.fixture
-def setup_workingdirectory(request):
+def setup_workingdirectory(qtbot, request):
     """Setup working directory plugin."""
     CONF.reset_to_defaults()
     use_startup_wdir = request.node.get_closest_marker('use_startup_wdir')

